### PR TITLE
[SPARK-19228][SQL] Migrate on Java 8 time from FastDateFormat for meet the ISO8601

### PR DESCRIPTION
**Pull request by sergey-rubtsov**
## What changes were proposed in this pull request?

Add support for inferring DateType and custom "dateFormat" option.
Fix an old bug with parse string to SQL's timestamp value in microseconds accuracy.
Add a type-widening rule in findTightestCommonType between DateType and TimestampType.

## How was this patch tested?

Fix some tests to accord with an internationally agreed way to represent dates.
Add an end-to-end test case and unit tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala [hunk 1]
```diff
@@ -107,20 +107,26 @@ class UnivocityParserSuite extends SparkFunSuite {
     assert(parser.makeConverter("_1", BooleanType, options = options).apply("true") == true)
 
     val timestampsOptions =
-      new CSVOptions(Map("timestampFormat" -> "dd/MM/yyyy hh:mm"), "GMT")
+      new CSVOptions(Map("timestampFormat" -> "dd/MM/yyyy HH:mm"), "GMT")
```
Comment (sergey-rubtsov) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala]: In accordind to official documentation, this test must not pass, because "hh" for hours means that hour in am/pm (1-12):
https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html
https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala [hunk 2]
```diff
@@ -107,20 +107,26 @@ class UnivocityParserSuite extends SparkFunSuite {
     assert(parser.makeConverter("_1", BooleanType, options = options).apply("true") == true)
 
     val timestampsOptions =
-      new CSVOptions(Map("timestampFormat" -> "dd/MM/yyyy hh:mm"), "GMT")
```
Comment (sergey-rubtsov) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala]: In accordind to official documentation, this test must not pass, because "hh" for hours means that hour in am/pm (1-12):
https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html
https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html

**Comment by sergey-rubtsov** [comment]
cc @viirya @mgaido91 @dongjoon-hyun

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
@sergey-rubtsov, would we be able to add a configuration to control this behaviour? Sounds we should better have a configuration to control this behaviour for now since the date / timestamp parsing logic is affected.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala [hunk 3]
```diff
@@ -107,20 +107,26 @@ class UnivocityParserSuite extends SparkFunSuite {
...
-    val customDate = "31/01/2015"
     val dateOptions = new CSVOptions(Map("dateFormat" -> "dd/MM/yyyy"), "GMT")
-    val expectedDate = dateOptions.dateFormat.parse(customDate).getTime
+    val customDate = "31/01/2015"
+
+    val expectedDate = LocalDate.parse(customDate, dateOptions.dateFormatter)
+      .atStartOfDay(options.timeZone.toZoneId)
+      .toInstant.toEpochMilli
     val castedDate =
-      parser.makeConverter("_1", DateType, nullable = true, options = dateOptions)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala]: I would keep this line as was.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 4]
```diff
@@ -119,7 +119,6 @@ class CSVOptions(
   val positiveInf = parameters.getOrElse("positiveInf", "Inf")
   val negativeInf = parameters.getOrElse("negativeInf", "-Inf")
 
-
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: Sounds unrelated change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 5]
```diff
@@ -143,6 +145,12 @@ object DateTimeUtils {
     millisLocal - getOffsetFromLocalMillis(millisLocal, timeZone)
   }
 
+  def dateTimeToMicroseconds(localDateTime: LocalDateTime, timeZone: TimeZone): Long = {
+    val microOfSecond = localDateTime.getLong(ChronoField.MICRO_OF_SECOND)
+    val epochSecond = localDateTime.atZone(timeZone.toZoneId).toInstant.getEpochSecond
+    epochSecond * 1000000L + microOfSecond
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: `1000000L` -> `MICROS_PER_SECOND`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 6]
```diff
@@ -140,14 +141,23 @@ private[csv] object CSVInferSchema {
   private def tryParseDouble(field: String, options: CSVOptions): DataType = {
     if ((allCatch opt field.toDouble).isDefined || isInfOrNan(field, options)) {
       DoubleType
+    } else {
+      tryParseDate(field, options)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: Is this a behavior change? Previously timestamp type, now date type/timestamp type?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchemaSuite.scala [hunk 7]
```diff
@@ -59,13 +59,21 @@ class CSVInferSchemaSuite extends SparkFunSuite {
     assert(CSVInferSchema.inferField(IntegerType, textValueOne, options) == expectedTypeOne)
   }
 
-  test("Timestamp field types are inferred correctly via custom data format") {
-    var options = new CSVOptions(Map("timestampFormat" -> "yyyy-mm"), "GMT")
+  test("Timestamp field types are inferred correctly via custom date format") {
+    var options = new CSVOptions(Map("timestampFormat" -> "yyyy-MM"), "GMT")
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchemaSuite.scala]: Why we need to change this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 8]
```diff
@@ -90,6 +90,7 @@ private[csv] object CSVInferSchema {
           // DecimalTypes have different precisions and scales, so we try to find the common type.
           findTightestCommonType(typeSoFar, tryParseDecimal(field, options)).getOrElse(StringType)
         case DoubleType => tryParseDouble(field, options)
+        case DateType => tryParseDate(field, options)
```
Comment (mgaido91) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: this also is a behavior change. Shall we document it?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchemaSuite.scala (hunk 7 above)
Comment (sergey-rubtsov) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchemaSuite.scala]: "yyyy-mm" means years and minutes, this is date format, this is time format
yyyy-MM" means years and months, but I do not insist on this change

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 8 above)
Comment (sergey-rubtsov) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: I can do it, but where exactly it should be documented?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala (hunk 3 above)
Comment (sergey-rubtsov) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala]: ok

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 6 above)
Comment (sergey-rubtsov) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: For example, by mistake we have identical "timestampFormat" and "dateFormat" options.
Let it be "yyyy-MM-dd"
'TimestampType' (8 bytes) is larger than 'DateType' (4 bytes)
So if they can overlap, we need to try parse it as date firstly, because both of these types are suitable, but you need to try to use a more compact by default and it will be correct inferring of type

**Comment by sergey-rubtsov** [comment]
@HyukjinKwon please, clarify, how to add a configuration to control this behaviour? Do you mean to keep backward compatibility? 
But it seems, now, we don't handle a Spark "DataType" at all. 
I can change the behaviour in next way: 
Hanlde "DataType" if option "dateFormat" was set by customer, if it is not set, ignore it at all, is it okay?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 5 above)
Comment (sergey-rubtsov) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: thanks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 6 above)
Comment (sergey-rubtsov) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: At the moment, DateType here is ignored at all, I'm not sure that it was conceived when the type was created

**Comment by mgaido91** [comment]
@sergey-rubtsov we have to keep backward compatibility. If a user upgrades, with your change a running application may break because of data not being anymore timestamp but date. We can add a new entry in `SQLConf` for that.

Moreover, we have to mention in the migration guide all the behavioral changes we introduce.

**Comment by HyukjinKwon** [comment]
ok to test

**Review by srowen — COMMENTED** [review]
I agree we can probably use Java 8 Date classes now instead of 3rd party ones, but can we do so without a behavior change? is it actually required?

**Comment by HyukjinKwon** [comment]
Looks difficult because the behaviours themselves are different. One possibility is a fallback and the other possibility is configuration.

**Comment by HyukjinKwon** [comment]
I am going to suggest to close this since it's being active more then few weeks. It should be good to fix. Let me leave some cc's who might be interested in this just FYI. Feel free to take over this when you guys find some time or are interested in this. 

Adding @xuanyuanking, @mgaido91, @viirya, @MaxGekk, @softmanu who I could think of for now. Feel free to ignore my cc if you guys are busy or having more important fixes you guys are working on.

**Comment by MaxGekk** [comment]
@xuanyuanking Are you still working on this? If not, I could continue.

**Comment by srowen** [comment]
@MaxGekk now that your change is merge, can this proceed, @xuanyuanking ? or is it obsolete?

**Comment by MaxGekk** [comment]
@srowen There is another PR with related changes - inferring `DateType` from CSV: https://github.com/apache/spark/pull/23202

**Comment by HyukjinKwon** [comment]
I'm going to close this since @sergey-rubtsov is inactive.

Most of feasible changes extracted from here are merged. One last part is date type inference which's going to need some more discussions (IMHO).
